### DISCUSSION TITLE
Example for Dynamic Cli Construction?

### DISCUSSION THREAD

**Discussion by jonathanlocke** [spectreconsole/spectre.console#1415] [Q&A] [answered]
Is there an example somewhere of how to dynamically populate a CommandApp with branches and commands? My issue is that I have several independent modules where each module would like to contribute branches and commands to the same command app. Some modules would also like to add branches and commands on to the same branch(es) as another module. Ideas? Thanks!

P.S. Hopefully there's a solution, but it would be interesting to support this sort of use with a method like MountCommand(path, command) where path would be a series of branches separated by slashes. For example:

```
app.MountCommand("git/new/bugfix", typeof(GitNewBugFix))
app.MountCommand("git/new/feature, typeof(GitNewFeature))
app.MountCommand("git/reset-all", typeof(GitResetAll))
```

We did this in Apache Wicket, and it turned out to be a winner in that case. -- Jon

**Comment by patriksvensson** [comment, marked as answer]
@jonathanlocke This is not a scenario that we support OOTB (and probably won't either), but you should be able to create something using the `IUnsafeConfigurator`. 

We haven't documented this, but take a look at the tests over at https://github.com/spectreconsole/spectre.console/blob/main/test/Spectre.Console.Cli.Tests/Unit/CommandAppTests.Unsafe.cs for more information.

**Reply by jonathanlocke** [reply]
Hi Partrik, I was able to use the existing type-safe API to accomplish this. I need to ask if I can contribute the full source first (if anyone wants it), but the basic API use looks like this code from the unit test:
```
            var application = new CommandApp();
            var tree = new CommandTree(application);
            tree.MountCommand<X, X.Settings>("a/b/c");
            tree.MountCommand<X, X.Settings>("a/b/d");
            tree.MountCommand<X, X.Settings>("a/c");
```
It turns out you just need to get creative with recursion and generics (by binding generic method parameters with type parameters).
